### PR TITLE
Rustdoc ty consistency fixes

### PR DESCRIPTION
**Pull request by CraftSpider**
Changes to make rustdoc cleaning of ty more consistent with hir, and hopefully use it in more places.

r? @camelid

### CODE REVIEW

**Comment by camelid** [comment]
Can you squash the "Format" commit into the previous commit?

**Comment by camelid** [comment]
I reviewed the first few commits. I'll try to review the rest soon.

**Comment by CraftSpider** [comment]
Format squashed in

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -890,10 +890,10 @@ fn clean_fn_decl_with_args(
 
 fn clean_fn_decl_from_did_and_sig(
     cx: &mut DocContext<'_>,
-    did: DefId,
+    did: Option<DefId>,
     sig: ty::PolyFnSig<'_>,
 ) -> FnDecl {
-    let mut names = if did.is_local() { &[] } else { cx.tcx.fn_arg_names(did) }.iter();
+    let mut names = did.map_or(&[] as &[_], |did| cx.tcx.fn_arg_names(did)).iter();
```
Comment (camelid) [inline src/librustdoc/clean/mod.rs:896]: Hmm, could this be a behavior change when this is called with an external function? I think if it is, it'd be a bugfix, but I wanted to check. In the inline code path (`inline.rs`), this is called with `Some(did)`, where `did` should be non-local.

**Review by camelid — COMMENTED** [review]
This looks good to me overall but I'm a bit unclear about d90138bec8f406b5c01c6bed4c0e597df2f3b00f. Could you explain the goal of that commit? Also, the last commit doesn't actually make hir and ty use the same code paths so the commit message is a little confusing. Most of the code is actually still different. It'd probably be hard to fully merge them until type aliases are represented in ty though.

**Comment by camelid** [comment]
Sorry for taking so long to review this by the way!

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (CraftSpider) [reply src/librustdoc/clean/mod.rs:896]: This should only change code on local functions, external functions always took the else case and are thus equivalent to `Some`. Local function behavior should be affected as an intentional fix related to this PR.

**Comment by CraftSpider** [comment]
Commit https://github.com/rust-lang/rust/commit/d90138bec8f406b5c01c6bed4c0e597df2f3b00f is reverting a change I made in an earlier PR that was necessary before these fixes - There's a JSON test that ensures removing it here didn't cause a regression. The comment about 'same code path' is meant to mean that the two different functions calculate what items are impl items the same way, not that they're fully the same code paths but that they should now use the same rules for what is an 'impl of a trait'.

**Review by camelid — APPROVED** [review]
Sorry again this took so long!

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/clean/mod.rs:896]: You're right, I got confused. Thanks!

**Pull request merged**